stdlib/src/collections/string.mojo

```diff
@@ -1090,19 +1090,34 @@ struct String(
             return self
         var self_len = self.byte_length()
         var other_len = other.byte_length()
-        var total_len = self_len + other_len
-        var buffer = Self._buffer_type()
-        buffer.resize(total_len + 1, 0)
-        memcpy(
-            buffer.data,
-            self.unsafe_ptr(),
-            self_len,
-        )
-        memcpy(
-            buffer.data + self_len,
-            other.unsafe_ptr(),
-            other_len + 1,  # Also copy the terminator
-        )
+        var buffer = Self._buffer_type(capacity=self_len + other_len + 1)
+        var ptr = buffer.unsafe_ptr()
+        memcpy(ptr, self.unsafe_ptr(), self_len)
+        memcpy(ptr + self_len, other.unsafe_ptr(), other_len + 1)
+        buffer.size = self_len + other_len + 1
+        return Self(buffer^)
+
+    fn __add__(self, other: StringSlice) -> String:
+        """Creates a string by appending a string slice at the end.
+
+        Args:
+            other: The string slice to append.
+
+        Returns:
+            The new constructed string.
+        """
+        if not self:
```

Comment:
**Suggestion** The logic here, in `__radd__`, and in the above `__add__` is all similar (get a pointer to the underlying data, do some memcpy of data, etc.).  Can we factor that into a private helper function and use it in these places to avoid duplication?